desc/objectives/_omnigenity.py

```diff
@@ -847,6 +846,379 @@ def _compute_B_eta_alpha(theta_B, zeta_B, B_mn):
+        normalize_target=True,
+        loss_function=None,
+        eq_grid=None,
+        field_grid=None,
+        deriv_mode="rev",
+        M_booz=None,
+        N_booz=None,
+        eq_fixed=False,
+        field_fixed=False,
+        overlap_penalty=1.0,
+        p=10,
+        name="Piecewise omnigenity",
+    ):
+        if target is None and bounds is None:
+            target = 0
+
+        self._eq = eq
+        self._field = field
+        self._eq_grid = eq_grid
+        self._field_grid = field_grid
+        self.M_booz = M_booz
+        self.N_booz = N_booz
+        self._eq_fixed = eq_fixed
+        self._field_fixed = field_fixed
+        self._overlap_penalty = overlap_penalty
+        self._p = p
+
+        if not eq_fixed and not field_fixed:
+            things = [eq, field]
+        elif eq_fixed and not field_fixed:  # Need to test this
+            things = [field]
+        elif field_fixed and not eq_fixed:
+            things = [eq, field]
+        else:
+            raise ValueError("Cannot fix both the eq and field.")
+
+        super().__init__(
+            things=things,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        if self._eq_fixed:
+            eq = self._eq
+            field = self.things[0]
+        elif self._field_fixed:
+            eq = self.things[0]
+            field = self._field
+        else:
+            eq = self.things[0]
+            field = self.things[1]
+
+        M_booz = self.M_booz or 2 * eq.M
+        N_booz = self.N_booz or 2 * eq.N
+
+        # default grids
+        if self._eq_grid is None:
+            rho = 0.5
+            eq_grid = LinearGrid(
+                rho=rho, M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, sym=False
+            )
+        else:
+            eq_grid = self._eq_grid
+
+        self._dim_f = eq_grid.num_nodes
+        self._eq_data_keys = ["|B|", "theta_B", "zeta_B"]
+        self._field_data_keys = ["|B|_pwO", "Q_pwO"]
+
+        errorif(eq_grid.sym, msg="eq_grid must not be symmetric")
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        profiles = get_profiles(self._eq_data_keys, obj=eq, grid=eq_grid)
+        eq_transforms = get_transforms(
+            self._eq_data_keys,
+            obj=eq,
+            grid=eq_grid,
+        )
+
+        # we need a uniform grid to get correct surface averages for iota
```

Comment:
iirc the boozer transform also needs a uniform grid so you can use the same one for both (and add a check that the eq grid is uniform)